[쇼핑 주문 1, 2 단계] 페토 미션 제출합니다.

chanho0908 opened the pull request:
# 셀프 체크리스트

## [이번 미션 커밋만 모아놓은 커밋입니다](https://github.com/woowacourse/android-shopping-order/pull/100/files/0268b70f516180f51bc6735c1152a610b91560e9..d448684392f786e377807defab5d09e618ce413a)

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [x] 프로그램이 정상적으로 작동하는가?
- [ ] 모든 테스트가 통과하는가?
- [ ] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [ ] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

 <!--
 학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
 정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
 -->
 

## 코드 리뷰 커뮤니케이션

### 안녕하세요 말리빈 ! 

마지막 미션의 리뷰이 페토입니다 ☺️

제가 이번 미션 페어 기간동안 테코톡 일정이 겹치게 되어 일정 조정에 실패해 미션을 다 완수하지 못했습니다 😭

에러처리, 테스트 코드도 작성하지 못했네요 🥲

말리빈과 꼭 리뷰어로 만난뵙고 싶었던 만큼 적어도 완성을 했어야 했는데 미션을 완수하기엔 제가 아직 부족합니다 

최대한 빠른 시간내에 구현하여 재요청 드리겠습니다! 🫡

### 서버 통신을 위한 JSON 직렬화 라이브러리를 선택한 이유
저희는 `kotlinx-serialization`을 선택했습니다. 

우선 `Kotlin` 언어 차원에서 제공하는 라이브러리다보니 외부 라이브러리에 비해 안정감있다고 생각했어요 !

또한 Gson은 리플렉션 기반으로 동작한다는 점에서 감점이었습니다.

아직 리플렉션에 대한 이해가 부족해 깊은 분석은 못했지만,

제가 느끼기에 리플렉션은 할 수 없는 것을 강제로 할 수 있게 만드는 것 같아요

그렇다면 당연히 그에 따른 문제가 발생할 여지가 있고요 ! 

이러한 이유 때문에 `kotlinx-serialization`을 선택했습니다. 

📌 [GitHub에서 PR에 댓글 남기는 방법](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request)

## 참고 자료

 <!-- 관련 문서나 참고할 만한 자료를 추가하세요. -->

## 스크린샷

 <!-- 필요하다면 스크린샷을 첨부하세요. -->

https://github.com/user-attachments/assets/484cd6b3-c3ab-477a-8487-b022bb896f68


![]()

chanho0908 committed:
git squash complete

Commit:
chore: 불필요한 코드 제거

chanho0908 committed:
build: add shimmer dependency

chanho0908 committed:
docs: 스켈레톤 UI 노출 기능 요구 사항 추가

chanho0908 committed:
feat: shimmer placeholder UI 추가

- shimmer_background 색상 추가
- shimmer 효과를 위한 item_shimmer_product.xml 레이아웃 추가

chanho0908 committed:
feat: 메인 화면 로딩 중 스켈레톤 UI 표시

메인 화면에서 데이터를 불러오는 동안 스켈레톤 UI를 표시하도록 기능을 추가했습니다.
- `MainViewModel`에 로딩 상태를 관리하는 `_isLoading` LiveData를 추가했습니다.
- `loadInitial` 함수 호출 시 로딩 상태를 true로 변경하고, 데이터 로드 완료 후 false로 변경합니다.
- `MainActivity`에서 `isLoading` LiveData를 관찰하여 로딩 상태에 따라 스켈레톤 UI와 상품 목록 RecyclerView의 가시성을 조절합니다.
- `activity_main.xml`에 ShimmerFrameLayout을 활용한 스켈레톤 UI를 추가했습니다.

Commit:
docs: 서버 연동 기능 요구 사항 추가

Commit:
build: Retrofit, kotlinx serialization 의존성 추가

Commit:
fix: 호환성 버전 해결

Commit:
feat: 상품 목록 네트워크 응답 모델 추가

Commit:
feat: 서버 상품 목록 불러오기 기능 추가

chanho0908 committed:
build: add logging interceptor dependency

chanho0908 committed:
feat: add http logging interceptor

chanho0908 committed:
refactor: Simplify product loading logic in MainViewModel

chanho0908 committed:
refactor: Remove unused ProductWithCartLoader from MainViewModelFactory

chanho0908 committed:
refactor(MainViewModel): 상품 목록 호출 로직 코드 최적화

chanho0908 committed:
feat: BasicAuthentication Interceptor 추가

chanho0908 committed:
feat: 장바구니에 상품을 추가하는 기능 구현

chanho0908 committed:
refactor: 메인 화면 상품 추가 시 서버 연동 로직 추가

- `ProductState`에 `increaseCartQuantity2` 함수를 추가하여 장바구니 수량 증가 시 서버와 동기화하도록 변경했습니다.
- `MainViewModel`에서 상품 추가 시 `cartRepository.addCart`를 호출하여 서버에 장바구니 정보를 전달합니다.
- `MainViewModelFactory`에서 `DefaultCartRepository`를 주입받도록 수정했습니다.
- `MainActivity`에서 `ViewModelProvider.Factory`에 `DefaultCartRepository`를 전달하도록 변경했습니다.
- `ProductUiState`의 `canIncreaseCartQuantity` 함수를 `increaseCartQuantity`로 변경하고, 반환 타입을 `ProductState`로 수정했습니다.

chanho0908 committed:
refactor: AppContainer에 새로운 CartRepository 의존성 추가

chanho0908 committed:
refactor: data 계층의 network 모델을 entity에서 response로 변경

chanho0908 committed:
feat: 장바구니 목록 조회 응답 DTO 추가

장바구니 목록 조회 API 응답을 위한 데이터 클래스들을 추가합니다.
- `CartsResponse`: 페이지네이션 정보를 포함한 전체 응답
- `Content`: 장바구니 상품 정보
- `Product`: 상품 상세 정보
- `Pageable`: 페이지네이션 관련 정보
- `SortX`: 정렬 관련 정보

Commit:
feat: `ProductState`에 `cartId` 추가

Commit:
fix: API 응답 DTO에 Serializable 어노테이션 추가

Commit:
feat: 장바구니 상품 목록 조회 API 응답 모델 정의

Commit:
feat: 장바구니 목록 조회 API 연동

Commit:
refactor: 상품 목록 로드 시 장바구니 정보 병합

상품 목록을 불러올 때 장바구니에 담긴 상품의 수량 정보를 함께 가져와 UI에 반영하도록 수정했습니다.

- `loadNextPage` 함수에서 상품 정보 로드 후 `loadCartAndMerge` 함수를 호출합니다.
- `loadCartAndMerge` 함수는 장바구니 정보를 로드하여 `mergeProductWithCart` 함수를 통해 상품 정보와 병합합니다.
- `mergeProductWithCart` 함수는 각 상품에 대해 장바구니 정보를 찾아 `ProductState`에 반영합니다.
- 로딩 상태 관리 및 오류 로깅 방식을 개선했습니다.

chanho0908 committed:
feat: 장바구니 상품 수량 수정 API 연동

chanho0908 committed:
refactor: 상품 및 장바구니 정보 로드 로직 개선

상품 목록과 장바구니 목록을 동시에 로드하고 병합하도록 로직을 변경했습니다.

- `loadNextPage` 함수를 `loadProductsAndCarts`로 변경하여 상품과 장바구니 정보를 함께 처리합니다.
- `loadCartAndMerge` 함수를 `mergeProductsWithCarts`로 변경하고, 상품과 장바구니 정보를 병합하여 `ProductState` 목록을 생성합니다.
- 로딩 상태 관리 및 오류 처리 방식을 개선하여 `setLoading` 및 `handleError` 함수를 사용합니다.
- 장바구니 수량 증가 시, 이미 장바구니에 담긴 상품이면 수량 업데이트 API를, 아니면 추가 API를 호출하도록 수정했습니다.

chanho0908 committed:
feat: 장바구니 상품 삭제 기능 구현

장바구니에 담긴 특정 상품을 삭제하는 기능을 추가합니다.

chanho0908 committed:
feat: 상품 목록 화면 장바구니 삭제 기능 추가

chanho0908 committed:
refactor: 수량 변경 컴포넌트 데이터 바인딩 방식 변경

수량 변경 컴포넌트(`item_quantity_changer.xml`)에서 `ProductState` 모델 전체를 받는 대신, 필요한 데이터(`cartId`, `quantity`, `containerVisibility`)만 각각 받도록 수정했습니다.

- `item_quantity_changer.xml`:
    - `model` 변수를 제거하고 `cartId`, `quantity`, `containerVisibility` 변수를 추가했습니다.
    - 이벤트 핸들러 및 텍스트, 가시성 설정 시 새로운 변수를 사용하도록 수정했습니다.
- `item_cart.xml`, `item_product.xml`:
    - `item_quantity_changer` 레이아웃을 include 할 때, `bind:model` 대신 새로운 변수들을 개별적으로 바인딩하도록 수정했습니다.
    - `item_cart.xml`: 모델 타입을 `ProductState`에서 `CartState`로 변경하고, `model`의 필드를 직접 참조하도록 수정했습니다.
- `activity_detail.xml`:
    - `item_quantity_changer`를 include 하는 부분을 주석 처리했습니다. (추후 기능 구현 시 재활성화 예정)

chanho0908 committed:
feat: `CartState` 및 `ShoppingCart`를 `CartState`로 변환하는 확장 함수 추가

chanho0908 committed:
feat: 장바구니 화면 상품 추가,삭제, 페이징 기능 API 연동 완료

Commit:
fix: 장바구니에 담긴 상품 수량 감소 시 cartId가 null로 초기화되지 않는 버그 수정

Commit:
feat: 상세 화면 상품 정보 API 연동

chanho0908 committed:
refactor: `ProductState` 및 관련 도메인 모델 수정

- `IncreaseState` 클래스를 삭제하고, 수량 증가 가능 여부 판단 로직을 `ProductState` 내부 또는 사용하는 쪽으로 이동시킵니다.
- `Quantity` 클래스에서 `isExceeded` 함수를 제거합니다. 수량 초과 여부 판단은 해당 로직이 필요한 곳에서 직접 처리하도록 변경합니다.
- `Product` 도메인 클래스에서 `canIncrease` 함수를 제거합니다. `IncreaseState` 삭제에 따른 변경입니다.
- `ProductsDataSource2`의 불필요한 주석을 제거합니다.

chanho0908 committed:
refactor: 상품 상세 화면 수량 변경 로직 수정 및 ViewModel 의존성 정리

- `DetailViewModelFactory` 및 `DetailViewModel`에서 불필요한 `productRepository` 및 `cartRepository` 의존성을 제거했습니다.
- `ProductState`에서 `increaseCartQuantity` 함수의 분기를 제거하고 단순 증가 로직으로 변경했습니다.
- `DetailUiState`에 `cartQuantity`와 `productId` 프로퍼티를 추가하여 데이터 접근을 용이하게 했습니다.
- 상품 상세 화면(`activity_detail.xml`)에서 수량 변경 UI가 ViewModel의 상태를 올바르게 반영하도록 수정했습니다.
- `DetailViewModel`의 `saveCart` 로직을 `DefaultCartRepository`를 사용하도록 변경했습니다.
- `ProductUiState`의 `increaseCartQuantity` 함수 호출을 새로운 `increaseCartQuantity`로 변경했습니다.

chanho0908 committed:
feat: 장바구니 상품 추가 시 ID 반환

장바구니에 상품을 추가하는 `addCart` 함수의 콜백에서 응답 헤더의 Location 값으로부터 cartId를 추출하여 반환하도록 수정했습니다.

chanho0908 committed:
feat: 장바구니 목록 쉬머 아이템 레이아웃 추가

장바구니 목록 로딩 중에 표시될 쉬머 효과를 위한 `item_shimmer_cart.xml` 레이아웃 파일을 추가합니다.

- 상품명과 상품 이미지를 위한 `View` 요소를 포함합니다.
- 배경색과 테두리는 기존 장바구니 아이템과 유사하게 설정합니다.
- 각 `View`의 배경색은 쉬머 효과를 위해 `@color/shimmer_background`로 설정합니다.

chanho0908 committed:
refactor: HistoryLoader 로직 변경

- HistoryLoader에서 `ProductRepository` 대신 `DefaultProductRepository`를 사용하도록 변경했습니다.
- 여러 상품 ID에 대해 `loadProduct`를 호출하고 결과를 취합하여 `HistoryState` 목록을 생성합니다.
- 각 상품 로드 시 발생할 수 있는 오류를 처리하고, 모든 상품 로드가 완료되거나 오류 발생 시 콜백을 호출합니다.
- 콜백은 `Result` 타입으로 변경되어 성공 또는 실패 결과를 전달합니다.

chanho0908 committed:
refactor: CartRepositoryImpl 스레드 제거 및 getCartSinglePage 파라미터 타입 변경

- `CartRepositoryImpl`에서 `thread`를 사용하여 비동기 처리하던 로직을 제거했습니다.
- `CartService`의 `getCartSinglePage` 메서드와 `CartDataSource2`의 `singlePage` 메서드의 `page`, `size` 파라미터 타입을 `Int`에서 `Int?`로 변경하여 선택적 값으로 수정했습니다.

chanho0908 committed:
refactor: `DefaultCartRepository.loadSinglePage` 매개변수 nullable로 변경

`loadSinglePage` 함수의 `page`와 `pageSize` 매개변수를 nullable 타입으로 변경하여, 기본값을 사용하거나 특정 페이지 정보를 전달하지 않을 수 있도록 유연성을 높였습니다.

chanho0908 committed:
feat: 상품 상세 화면에서 장바구니 담기 기능 수정

- 이미 장바구니에 담긴 상품을 또 담을 경우, 수량을 합산하여 업데이트하도록 변경
- 새로운 상품을 담을 경우, 기존 로직과 동일하게 장바구니에 추가

chanho0908 committed:
refactor: 코드 스타일 개선 및 불필요한 코드 제거

chanho0908 committed:
refactor: `ProductState` 및 `ShoppingCart`에 `productId` 접근자 추가

chanho0908 committed:
refactor: 장바구니 수량 동기화 로직 추가

chanho0908 committed:
refactor: DataSource 및 Repository 구조 변경

chanho0908 committed:
refactor: API 인증 정보 및 기본 URL을 `local.properties`에서 로드하도록 변경

chanho0908 committed:
docs: README 기능 요구사항 추가

- 장바구니에서 특정 상품만 골라 주문하는 기능 추가
- 상품 추천 기능 관련 요구사항 상세 명세

chanho0908 committed:
feat: 장바구니 상품 체크박스 기능 추가

장바구니 아이템 레이아웃에 체크박스를 추가하고, 체크 상태 변경 시 `onCheckedChanged` 핸들러를 호출하도록 구현했습니다.

- `item_cart.xml`: 체크박스 UI 추가 및 `onCheckedChanged` 이벤트 핸들러 바인딩
- `CartViewHolder.kt`: `onCheckedChanged` 핸들러 인터페이스 정의

chanho0908 committed:
feat: 장바구니 상품 선택 기능 추가

chanho0908 committed:
feat: 장바구니 화면 하단 주문 기능 및 Fragment 분리

- `CartActivity`의 화면 구성을 `FragmentContainerView`를 사용하도록 변경하여 `CartListFragment`를 포함시켰습니다.
- `CartListFragment`를 추가하여 장바구니 목록과 페이징 UI를 담당하도록 했습니다.
- `CartActivity` 하단에 전체 선택 체크박스, 총 주문 금액, 주문하기 버튼을 포함하는 레이아웃을 추가했습니다.
- `item_cart.xml`에서 수량 변경 UI의 레이아웃 제약을 수정했습니다.
- `strings.xml`에 주문하기 버튼 텍스트(`text_order`)를 추가했습니다.

chanho0908 committed:
fix: 수량 변경 UI 너비 수정

수량 변경 UI의 너비를 `wrap_content`에서 `match_parent`로 변경하여 부모 레이아웃의 너비를 채우도록 수정합니다.

chanho0908 committed:
refactor: CartUiState에 totalPrice, checkedProductCount, setAllItemsChecked 추가

- `totalPrice`: 선택된 상품들의 총 가격을 계산하는 프로퍼티 추가
- `checkedProductCount`: 선택된 상품들의 총 수량을 계산하는 프로퍼티 추가
- `setAllItemsChecked`: 모든 상품의 선택 상태를 한 번에 변경하는 함수 추가

chanho0908 committed:
feat: 주문 완료 화면 레이아웃 추가

- `fragment_order_complete.xml`: 주문 완료 화면 레이아웃 파일 추가
  - 추천 상품 제목, 설명, RecyclerView 포함
- `strings.xml`: 주문 완료 화면에서 사용될 문자열 추가
  - `text_order_complete_recommend_title`: "이런 상품은 어떠세요?"
  - `text_order_complete_recommend_description`: "* 최근 본 상품 기반으로 좋아하실 것 같은 상품들을 추천해드려요."

chanho0908 committed:
feat: 장바구니 상품 페이징 기능 추가 및 전체 선택 상태 유지

- `CartAdapter`가 페이지 번호를 받아 해당 페이지의 아이템만 표시하도록 변경
- `CartViewModel`에 `historyLoader` 추가 및 `loadCarts()` 호출 시점 변경
- `CartViewModel`에서 장바구니 목록 로드 시 전체 선택 상태(`allChecked`)를 `CartState`에 반영
- `CartUiState`에 `allChecked` 프로퍼티 추가 및 `setAllItemsChecked` 호출 시 `allChecked` 상태도 함께 변경
- `CartListFragment`에서 `CartAdapter`에 페이지 번호를 전달하여 페이징된 아이템 목록을 표시하도록 수정

chanho0908 committed:
feat: 주문 완료 화면 추가 및 화면 전환 기능 구현

chanho0908 commented:
@coderabbitai review

chanho0908 committed:
refactor: CartViewHolder 팩토리 메서드 추가

CartViewHolder 생성을 위한 `of` 팩토리 메서드를 추가하여 뷰 바인딩 로직을 캡슐화하고 인스턴스 생성을 단순화합니다.

- `CartViewHolder.kt`: `of` 팩토리 메서드 추가
- `CartAdapter.kt`: `CartViewHolder.of`를 사용하여 ViewHolder 인스턴스 생성

malibinYun commented on app/build.gradle.kts:
```diff
@@ -1,11 +1,27 @@
+    val userId = properties.getProperty("user_id") ?: ""
+    val userPassword = properties.getProperty("user_password") ?: ""
+    val baseUrl = properties.getProperty("base_url") ?: ""
+
```
baseurl을 여기에 배치한 이유는 무엇인가요?

chanho0908 replied:
https://github.com/woowacourse/android-shopping-order/pull/100/commits/ef2098fae097097349d6ce8e004d493a4ce42683

해당 내용 관련 커밋입니다 !

이 부분은 순수하게 제가 인라인 방식으로 선언 가능한 것을 인지하지 못했습니다 🥲

malibinYun commented on app/src/main/java/woowacourse/shopping/data/datasource/CartDataSource.kt:
```diff
@@ -1,25 +1,131 @@
+                        callback(Result.success(cartId))
+                    } else {
+                        callback(Result.failure(NullPointerException()))
+                    }
```
실제로 요청이 실패하면 NullPointerException이 에러 콘솔에 아무 메시지 없이 나올 것이 예상되네요.
실제 무엇 때문에 실패했는지 디버깅할 수 있을까요?

또, 어디는 HttpException, 어디는 NullPointerException이네요.
어떤 의도를 담으셨나요?

chanho0908 replied:
### 관련 리뷰
https://github.com/woowacourse/android-shopping-order/pull/100#discussion_r2115131214 

https://github.com/woowacourse/android-shopping-order/pull/100#discussion_r2115139104

### 관련 커밋
https://github.com/woowacourse/android-shopping-order/pull/100/commits/37d99b07bf1fab4b492d4484451667f51aad7f77 https://github.com/woowacourse/android-shopping-order/pull/100/commits/f51f864356718fd1b937706cc467aa1233252196 https://github.com/woowacourse/android-shopping-order/pull/100/commits/e01cba75019d60485887faf1fe888ddd62a04b6f https://github.com/woowacourse/android-shopping-order/pull/100/commits/bad654a65308f7d9a66dc6f4e23ea1febe1a29e8

해당 내용 관련 커밋입니다 ! 

죄송스럽게도 정신없이 구현하다보니 큰 의미 없이 우다다다 구현했습니다 🥲

완성 후 리팩토링 예정이였는데 미션을 다 구현하지 못해 이 부분도 수정하지 못했습니다 !

리뷰 적용 과정에서 다음과 같은 수정이 이루어졌습니다.

> 실제로 요청이 실패하면 NullPointerException이 에러 콘솔에 아무 메시지 없이 나올 것이 예상되네요.
실제 무엇 때문에 실패했는지 디버깅할 수 있을까요?

리뷰 내용을 고민하며 어떻게 하면 에러 메시지를 통해 디버깅을 쉽게 할 수 있을지 고민했습니다 !
request 객체의 내부 구현을 살펴보니 다음과 같이 구성되어 있던 것을 확인할 수 있었습니다.
<img width="459" alt="image" src="https://github.com/user-attachments/assets/223125ee-b1ba-4159-a905-8f0a62108a4f" />

+ url : 요청 API
+ method : HTTP Method
+ headers : HTTP 요청에 포함될 헤더
+ body : 요청 본문 

이걸 조합해서 다음과 같은 메시지를 조합해봤습니다 !

> E/API: Missing body: GET http://techcourse-lv2-alb-974870821.ap-northeast-2.elb.amazonaws.com(200 OK)

API 호출 관련 로직에 대한 에러처리를 통해 다음과 같이 에러 발생시 메시지를 보여주도록 수정했습니다 !

https://github.com/user-attachments/assets/17154231-02f0-4fa8-8f20-07e6b0ea7916

malibinYun commented on app/src/main/java/woowacourse/shopping/data/datasource/CartDataSource.kt:
```diff
@@ -1,25 +1,131 @@
+        cartId: Long,
+        quantity: Int,
+        callback: (Result<Unit?>) -> Unit,
+    ) {
```
응답 결과를 Unit? (nullable)으로 설정한 이유는 무엇인가요?

chanho0908 replied:
API 호출에 성공해 `response`가 성공적으로 반환되어도 반환 타입이 없는 경우 
`Body`가 비어있기 때문에 `nullable`로 설정했습니다.

https://github.com/woowacourse/android-shopping-order/pull/100#discussion_r2122482548

해당 코멘트에서 non-nullable 로 수정헸습니다 !

malibinYun commented on app/src/main/java/woowacourse/shopping/data/network/BasicAuthentication.kt:
```diff
@@ -0,0 +1,29 @@
+        return chain.proceed(newRequest)
+    }
+
+    private fun basicAuthenticationHeader(): String {
+        val userId = woowacourse.shopping.BuildConfig.USER_ID
+        val userPassword = woowacourse.shopping.BuildConfig.USER_PASSWORD
+
```
앱에 로그인 기능이 생긴다면, 이 부분은 어떤 변경점이 생겨날까요?
유연하게 대처가 가능한 형태일까요?

malibinYun commented on app/src/main/java/woowacourse/shopping/data/network/service/CartService.kt:
```diff
@@ -0,0 +1,36 @@
+import retrofit2.http.Query
+import woowacourse.shopping.data.network.request.CartItemRequest
+import woowacourse.shopping.data.network.response.carts.CartsResponse
+
+interface CartService {
```
서비스 객체를 하나로 만들지 않고 파일을 쪼갠 이유는 무엇인가요?

chanho0908 replied:
API에서 내려주는 도메인이 "상품"과 "장바구니"로 나누어져 있어 분리하게 되었습니다.

다만 이 부분에 대해서 다시 고민해보니 서버에서 이 두 도메인을 "상품"이라는 하나의 도메인으로 합치기로
수정한다면 클라이언트도 수정되어야 하는 것인가 ? 에 대한 의문이 드는 것 같네요 🤔

malibinYun replied:
Cart인 Product, Product인 Cart라면  어느 파일에 위치시켜야할까요?
도메인들이 점점 늘어나게된다면, 그 경우의 수들은 모두 어느 파일에 위치시켜야할까요?

malibinYun commented on app/src/main/java/woowacourse/shopping/data/repository/DefaultCartRepository.kt:
```diff
@@ -0,0 +1,97 @@
+                    } else {
+                        callback(Result.failure(NullPointerException()))
+                    }
+                },
```
data source에서 가공했던 Result 객체를 그대로 콜백에 넘기지 않고 여기서 다시 Result 객체를 재구성하는 이유는 무엇인가요?

malibinYun commented on app/src/main/java/woowacourse/shopping/view/cart/CartActivity.kt:
```diff
@@ -63,49 +57,50 @@ class CartActivity : AppCompatActivity(), CartQuantityHandler, CartAdapter.Handl
+                }
+            }
+            is OrderCompleteFragment -> {}
+        }
```
함수명만 보고는 그래서 이 함수가 무엇을 수행하는지는 구현을 반드시 보아야겠어요.
이름과 내용이 일치한다고 볼 수 있을까요?

물론 클릭 이벤트를 발생하는 곳에서 호출하는 쪽에서는 자연스러울 수 있겠으나, 
다른 클릭 이벤트가 아닌 곳에서 프래그먼트 이동에 대한 (함수 내용) 재활용을 원할 때, 이 함수를 사용한다면, 그것은 자연스럽게 보일까요?

chanho0908 replied:
https://github.com/woowacourse/android-shopping-order/pull/100/commits/de8a0cb5dc73dd61a8ef5e94414954a6599fbdc7

말씀해주신 것 처럼 이 함수가 가지는 **행위** 측면에서는 부적합하다고 생각해 수정했습니다 !

malibinYun commented on app/src/main/java/woowacourse/shopping/view/cart/CartListFragment.kt:
```diff
@@ -0,0 +1,92 @@
+        R.layout.fragment_cart_list,
+    ),
+    CartQuantityHandler,
+    CartAdapter.Handler {
```
CartListFragment 는 Fragment 이지요. (CartListFragment is a Fragment)
CartListFragment CartQuantityHandler인가요? 또, CartAdapter.Handler 인가요?

chanho0908 replied:
https://github.com/woowacourse/android-shopping-order/pull/100/commits/a2a2aeafd6c50bfdbda7e25e4dd0fbe90b6b82cb

해당 리뷰 관련 커밋입니다 !

이벤트를 트리거 시키는 주체가 ViewModel이기 때문에 ViewModel에 익명 객체로 수정했습니다 !

malibinYun commented on app/src/main/java/woowacourse/shopping/view/cart/adapter/RecommendAdapter.kt:
```diff
@@ -0,0 +1,34 @@
+        return ProductViewHolder(
+            parent,
+            handler as ProductViewHolder.Handler,
+            handler as CartQuantityHandler,
```
캐스팅에 실패한다면 앱이 크래시가 나겠어요.
사용자에게 좋은 사용성을 전달할 수 있을까요?

malibinYun commented on app/src/main/java/woowacourse/shopping/view/main/vm/MainViewModel.kt:
```diff
@@ -1,162 +1,202 @@
-        _uiEvent.setValue(event)
+    private fun setLoading(isLoading: Boolean) {
+        _isLoading.postValue(isLoading)
     }
```
여기에서는 postValue를, cartViewModel에서는 setValue를 쓰고 있네요.
각각 다르게 쓰는 이유는 무엇인가요?

chanho0908 replied:
https://github.com/woowacourse/android-shopping-order/pull/100/commits/befd21b484cc540270fec121e1075bebd658d3e4

해당 리뷰 관련 커밋입니다 !

이 부분도 정신없이 구현하다가 큰 의미없이 구현되었네요  🥲

malibinYun commented on app/src/main/java/woowacourse/shopping/view/main/vm/MainViewModel.kt:
```diff
@@ -1,162 +1,202 @@
     val uiEvent: SingleLiveData<MainUiEvent> get() = _uiEvent
 
+    private val _isLoading = MutableSingleLiveData(true)
+    val isLoading: SingleLiveData<Boolean> get() = _isLoading
```
isLoading은 uiState안에 존재하지 않는 이유는 무엇인가요?

malibinYun reviewed (changes requested):
1, 2단계 미션고생 많으셨어요.
커밋 범위를 잘 지정해주셔서 이번 변경사항만 제한해서 보고 피드백을 남길 수 있었어요 :)
여러 생각해볼 거리에 코멘트 남겨두었으니 충분히 고민해보세요.
피드백 반영 후 다시 리뷰 요청 부탁드려요.

malibinYun commented:
저를 꼭 만나보고싶었다구요..? 무서운 거 아니죠?
이래놓고 드나나 재방문때 안반겨주시면 서운해할거예요

---

> ### 서버 통신을 위한 JSON 직렬화 라이브러리를 선택한 이유
> 저희는 kotlinx-serialization을 선택했습니다.
우선 Kotlin 언어 차원에서 제공하는 라이브러리다보니 외부 라이브러리에 비해 안정감있다고 생각했어요 !
또한 Gson은 리플렉션 기반으로 동작한다는 점에서 감점이었습니다.
아직 리플렉션에 대한 이해가 부족해 깊은 분석은 못했지만,
제가 느끼기에 리플렉션은 할 수 없는 것을 강제로 할 수 있게 만드는 것 같아요
그렇다면 당연히 그에 따른 문제가 발생할 여지가 있고요 !
이러한 이유 때문에 kotlinx-serialization을 선택했습니다.

아래 남겨 드릴 피드백은 페토의 말이 틀렸으니 고쳐라 라는 의미가 아닌, 이런 의견들이 있을 때 당신은 얼마나 자신의 주장을 논리적 근거를 뒷받침해 수호할 수 있는가에 대해 더 집중해보시면 좋겠어요.

kotlinx-serialization은 다른 직렬화/역직렬화 라이브러리의 등장에 상대적으로 매우 후발 주자인데요, 그만큼의 기간 차이동안의 라이브러리 성숙도가 높지 않다고 생각해요. 다른 외부 라이브러리에 비해 안정감 있다고는 공감이 가지 않아요. 다른 라이브러리들에서 지원해주는 모든 기능과 이전 라이브러리에서 해결한 버그들은 kotlinx-serialization이 모두 해결된 상태라고 정말 볼 수 있나요? "안정적" 측면에서요.

Gson은 리플랙션으로 동작하는 것이 맞아요. `할 수 없는 것을 강제로 할 수 있게 만드는 것이 당연히 문제가 발생할 여지가 있다`고 해주셨는데요, Retrofit은 굉장히 많은 코드를 리플랙션으로 만들고 있어요. 그렇다면 Retrofit은 문제가 많을텐데, 구글에서 권장하는 라이브러리라는 게 모순이란 생각이 드네요.
게다가 Gson은 최근까지만 해도 구글의 권장 직렬화/역직렬화 라이브러리였는데요, 이전에 사용하던 앱들은 항상 문제가 있다고 받아들이기에는 공감이 가지 않아요.

chanho0908 committed:
feat: 추천 상품 UI 구현

- `item_recomment.xml`: 추천 상품 아이템 레이아웃 추가
- `RecommendViewHolder`: 추천 상품 아이템 ViewHolder 구현
- `RecommendAdapter`: 추천 상품 목록 Adapter 구현
- `RecommendDiffUtil`: 추천 상품 목록 DiffUtil 구현

chanho0908 committed:
feat: 상품 목록 화면 UI 이벤트 개선 및 장바구니 이동 로직 추가

chanho0908 committed:
feat: 쉬머 로딩 UI 개선

chanho0908 committed:
feat: 상품 조회 API에 카테고리 필터링 기능 추가

chanho0908 committed:
refactor: CartState, CartUiState, ShoppingCart 및 Quantity 모델 개선

chanho0908 committed:
refactor: 주문 완료 Fragment 이름 변경 및 구현

chanho0908 committed:
refactor: HistoryRepository 및 관련 클래스 수정

chanho0908 committed:
refactor: HistoryRepository 및 관련 클래스 수정

chanho0908 committed:
feat: 주문 완료 화면을 추천 상품 화면으로 변경

chanho0908 committed:
feat: 상품 상세 화면에서 장바구니 이동 시 카테고리 정보 전달

chanho0908 committed:
refactor: MainViewModel의 saveHistory 메서드 파라미터 변경

chanho0908 committed:
refactor: AppContainer 및 BasicAuthentication 코드 정리

chanho0908 committed:
refactor: onClick 핸들러 파라미터에서 category 제거

chanho0908 committed:
refactor: 최근 본 상품 관련 메서드명 변경 및 카테고리 추가

chanho0908 committed:
refactor: CartListFragment 패키지 이동 및 관련 코드 수정

Commit:
feat: 추천 상품 수량 조절 및 UI 상태 관리 기능 추가

Commit:
refactor: CartListFragment의 수량 변경 메서드 파라미터 productId에서 cartId로 변경

chanho0908 commented:
@coderabbitai resume
@coderabbitai review

chanho0908 committed:
refactor: 불필요한 Log import 제거

- `CartDataSource.kt`: `android.util.Log` import 제거
- `RecommendFragment.kt`: `android.util.Log` import 제거

chanho0908 committed:
refactor: 상품 조회 기록 저장 로직 변경 및 상품 상세 화면 종료 시 기록 동기화

- 상품 조회 기록 저장 시점을 상품 상세 화면 진입 시점으로 변경했습니다.
- 상품 상세 화면 종료 시 `MainActivity`에서 조회 기록을 동기화하도록 변경했습니다.
- `MainActivity`의 `onSelectProduct`, `onClickHistory`에서 `saveHistory` 대신 `handleNavigateDetailEvent`를 호출하도록 수정했습니다.

chanho0908 committed:
refactor: 이벤트 핸들러 ViewModel로 이동

chanho0908 committed:
refactor: 네트워크 응답 처리를 위한 확장 함수 추가

- `enqueueWithResult`: Retrofit Call 객체에 대한 일반적인 응답 처리를 위한 확장 함수를 추가합니다. 성공 시 응답 본문을, 실패 시 예외를 `Result`로 감싸 콜백에 전달합니다.
- `enqueueWithExtractHeaderLocationResult`: Retrofit Call 객체에 대한 응답 처리 중, 성공 시 응답 헤더의 "Location" 값을 추출하여 콜백에 전달하는 확장 함수를 추가합니다. 실패 시 예외를 `Result`로 감싸 콜백에 전달합니다.

chanho0908 committed:
refactor: MainViewModel 에러 처리 로직 수정

chanho0908 committed:
feat: 네트워크 오류 발생 시 토스트 메시지 표시

chanho0908 committed:
refactor: CartViewModel 네트워크 에러 처리 추가

chanho0908 committed:
refactor: 응답 DTO에 BaseResponse 인터페이스 적용

chanho0908 committed:
feat: API 콜백 핸들러 ApiCallbackHandler 추가

chanho0908 committed:
refactor: 네트워크 오류 메시지 구체화 및 UI 이벤트 변수명 변경

chanho0908 committed:
refactor: `isLoading` LiveData 업데이트 방식 변경

chanho0908 committed:
refactor: BuildConfig 필드 설정 방식 변경

chanho0908 committed:
refactor: ViewModel에서 로딩 상태를 LiveData 대신 UiState로 관리하도록 변경

chanho0908 committed:
refactor: Unit 응답을 처리하도록 ApiCallbackHandler 개선

- `enqueueWithResult` 함수에서 제네릭 타입 `T`가 `Unit`인 경우, `response.body()`가 `null`이더라도 `Result.success(Unit)`을 반환하도록 수정했습니다.

chanho0908 committed:
refactor: `onClickOrderButton` 함수명을 `changeFragment`로 변경

chanho0908 committed:
refactor: DefaultProductRepository 콜백 로직 간소화

`productDataSource`의 콜백 결과를 그대로 전달하도록 수정하여 중복되는 `fold` 로직을 제거했습니다.

chanho0908 committed:
feat : 상품 상태 클래스 팩토리 메서드 추가 및 적용

chanho0908 committed:
refactor: 메인 화면 상품 목록 동기화 수정 및 코드 개선

chanho0908 committed:
fix: 최근 본 상품 목록이 유지되지 않는 문제 수정

chanho0908 committed:
refactor: HistoryRepository.saveHistory()의 onResult 콜백 제거

chanho0908 committed:
refactor: ProductUiState의 isFetching 토글 함수명 변경

chanho0908 committed:
refactor: 뷰모델의 수량 변경 로직을 UI 상태 객체로 이동

chanho0908 committed:
refactor: 상세 화면의 장바구니 추가 로직 변경

장바구니 추가 로직 변경 사항
- 기존: CartRepository에서 모든 카트 아이템을 불러온 후, 상세화면에서 해당 아이템을 검색하여 추가/갱신 로직 실행
- 변경: CartsSinglePage 객체 내부에서 해당 아이템을 검색하도록 변경

chanho0908 committed:
refactor: 상세 화면 초기화 로직 분리

chanho0908 committed:
refactor: CartViewModel 코드 개선

malibinYun commented:
원활한 다음 단계 미션 진행을 위해 오늘 밤 중에는 머지하도록 할게요 :)

chanho0908 committed:
fix: 장바구니에서 상품 삭제시 수량이 동기화 되지 않는 문제 수정

chanho0908 committed:
test: `ApiCallbackHandler` 테스트를 추가, `DummyApi` 추가

chanho0908 reviewed (commented):
안녕하세요 말리빈 !

이번에 드나 스텝으로 참여하게 되서 뵙게 된다면 맨발로 뛰어가겠습니다 🫡

미 완성되었던 기능과 리팩토링을 중점으로 진행했습니다.

다만 제가 느끼기에 테스트를 작성하기가 너무 어려워 우선은 
테스트 작성을 하지 않았습니다..

AI를 사용해 만들 수 도 있었겠지만 그 마저도 제가 코드를 온전히 이해하지 못했기 때문에
그 코드를 그대로 사용하는 것은 의미가 없다고 생각했어요 🥲

대신 그 동안 무림 고수마냥 쏴대던 장풍들을 최대한 없애기 위해 프로덕션 코드에 집중해봤습니다 !

malibinYun commented on app/build.gradle.kts:
```diff
@@ -1,10 +1,35 @@
+            "USER_PASSWORD",
+            "\"${properties.getProperty("user_password")}\"",
+        )
+        buildConfigField(
+            "String",
+            "BASE_URL",
+            "\"${properties.getProperty("base_url")}\"",
+        )
+    }
```
https://github.com/woowacourse/android-shopping-order/pull/100#discussion_r2115128334
질문을 바꿔볼게요.
코틀린 파일의 상수로 선언하지 않고 여기에 선언한 이유는 무엇인가요?

둘은 어떤 차이가 있을까요?
이곳에 배치하면 어떤 특징을 가져갈 수 있을까요?

malibinYun commented on app/src/main/java/woowacourse/shopping/data/ApiCallbackHandler.kt:
지금 이 파일만 봐도 enqueue에 대한 함수 커스텀이 3개나 들어있네요.
프로젝트가 점점 커진다면, 과연 3개로 끝날까요?
다른 개발자는 무엇을 선택해서 사용야할지 한 번에 알아낼 수 있을까요?

malibinYun commented on app/src/main/java/woowacourse/shopping/data/datasource/CartDataSource.kt:
```diff
@@ -0,0 +1,33 @@
+    private val service: CartService,
+    private val handler: ApiCallbackHandler,
+) {
+    fun addCart(
+        request: CartItemRequest,
+        callback: (Result<String>) -> Unit,
+    ) = handler.enqueueWithExtractHeaderLocationResult(service.addCart(request), callback)
+
+    fun singlePage(
+        page: Int?,
+        size: Int?,
+        callback: (Result<CartsSinglePage>) -> Unit,
+    ) = handler.enqueueWithDomainTransform(service.getCartSinglePage(page, size), callback)
+
+    fun updateCartQuantity(
+        cartId: Long,
+        quantity: Int,
+        callback: (Result<Unit>) -> Unit,
+    ) = handler.enqueueWithResult(service.updateCart(cartId, quantity), callback)
+
+    fun deleteCart(
+        cartId: Long,
+        callback: (Result<Unit>) -> Unit,
+    ) = handler.enqueueWithResult(service.deleteCart(cartId), callback)
```
이 파일만 보고, 어떤 함수가 어떤 값을 반환하는지 알 수 있을까요?
내용물을 아는 사람에게는 아무런 값을 반환하지 않는 함수로 보이지만, 다른 개발자가 보았을 때도 동일하게 그렇게 보일지 고민해보시면 좋겠네요

malibinYun commented on app/src/main/java/woowacourse/shopping/data/ApiCallbackHandler.kt:
```diff
@@ -0,0 +1,148 @@
+
+                    if (response.isSuccessful && location != null) {
+                        callback(Result.success(location))
+                    } else if (response.isSuccessful) {
+                        logMissingHeader(call, response)
+                        callback(Result.failure(NullPointerException()))
```
로직으로만 보면 응답이 response.isSuccessful로 성공인데, Result.failure를 반환하는 이유는 잘 모르겠네요.
NPE인 이유는 무엇인가요?
실제로 이 Exception이 발생했을 때, 에러 메시지에 아무것도 나타나지 않는다면 디버깅하기 쉬운 코드일까요?

malibinYun commented on app/src/main/java/woowacourse/shopping/data/repository/DefaultCartRepository.kt:
```diff
@@ -0,0 +1,42 @@
+import woowacourse.shopping.domain.repository.CartRepository
+
+class DefaultCartRepository(
+    private val dataSource: CartDataSource,
+) : CartRepository {
```
DataSource가 인터페이스가 아니었군요.
Repository와 DataSource는 어떨 때 사용하는걸까요?
어느 원천에서 온 데이터인지 이 이름만 보고는 알기가 너무 어렵네요.

malibinYun commented on app/src/main/java/woowacourse/shopping/view/main/vm/MainViewModel.kt:
```diff
@@ -0,0 +1,202 @@
+        carts: List<ShoppingCart>,
+    ) {
+        val newProducts = generateProductStates(productPage, carts)
+        historyLoader { result ->
+            result.onSuccess {
+                historyItems.postValue(it)
```
https://github.com/woowacourse/android-shopping-order/pull/100#discussion_r2115146796
postValue를 setValue로 바꾸면 어떤일이 발생할까요?
무슨 차이가 있을까요?

historyItems에서 가져온 값으로 setValue를하면 문제가 생기는데요, productRepository에서 가져온 값은 setValue를 해도 문제가 없어요. 왜일까요?

malibinYun commented on app/src/test/java/woowacourse/shopping/view/CartViewModelTest.kt:
```diff
@@ -0,0 +1,124 @@
+            callback(CartSinglePage(dummyPages.getOrElse(page) { emptyList() }, hasNext))
+        }
+
+        every { cartRepository.delete(any(), any()) } answers {
+            val productId = firstArg<Long>()
+            val callback = secondArg<(() -> Unit)>()
+            dummyPages.forEach { page ->
+                page.filter { it.productId != productId }
+            }
+            dummyPages.filter { it.isNotEmpty() }
+            callback()
+        }
```
`every { A } answers { B }`에서, A를 실행하면 B 사이드이팩트가 발생한다고 볼수있는데요,
B 안에 필요하지 않은 내용이 많아보이네요. 그래서 실제로 실행시키는 것만 제한해서 넣어보는 건 어떨까요?
준비할 내용이 더 길어진다면, 다른 개발자들이 이 테스트를 보았을때 그래서 무엇이 의도인지 한 번에 알아차릴 수 있을까요?

malibinYun commented on app/src/test/java/woowacourse/shopping/view/CartViewModelTest.kt:
```diff
@@ -0,0 +1,124 @@
+
+        val state = viewModel.uiState.getOrAwaitValue()
+
+        assertAll(
+            { assertThat(state.items).hasSize(5) },
+            { assertThat(state.items.map { it.item.id }).containsExactly(1L, 2L, 3L, 4L, 5L) },
+            { assertThat(state.pageState.page).isEqualTo(1) },
+            { assertThat(state.pageState.nextPageEnabled).isTrue() },
+            { assertThat(state.pageState.previousPageEnabled).isFalse() },
+        )
```
state 객체를 하나 만들어서 비교하는 것과 어떤 차이가 있을까요?
uiState 객체 안에 필드가 하나 더 생긴다면, 이 테스트를 변경해야하는지 알아차릴 수 있을까요?

malibinYun reviewed (approved):
1, 2단계 고생 많으셨어요.
라이브러리 선택에 관련한 내용에도 논리적 근거를 잘 찾아 답변해보시면 좋겠네요!
다음 미션 진행을 위해 이만 머지할게요.
피드백은 다음 미션에 반영 부탁드려요 :)
